Conceptual_Guide/Part_7-iterative_scheduling/model_repository/simple-gpt2/1/model.py

```diff
@@ -0,0 +1,197 @@
+            mapping (list): A list of indices that map the input tensors to the requests.
+        """
+
+        input_ids = []
+        mapping = []
+        for index, state in self.state.items():
+            if state != []:
+                mapping.append(index)
+                input_ids.append(state)
+
+        if not mapping:
+            return None, None, mapping
+
+        # Find the max sequence length
+        max_len = max([len(x) for x in input_ids])
+
+        # Pad the input tensors.
+        input_ids_torch = torch.cat(
+            [
+                torch.cat(
+                    [
+                        torch.tensor(
+                            [self.tokenizer.eos_token_id] * (max_len - len(x)),
+                            device=self.device,
+                        )
+                    ]
+                    + [x]
+                ).unsqueeze(0)
+                for x in input_ids
+            ]
+        )
+        attention_mask = torch.cat(
+            [
+                torch.cat(
+                    [
+                        torch.tensor([0] * (max_len - x.numel())),
+                        torch.tensor([1] * x.numel()),
+                    ]
+                ).unsqueeze(0)
+                for x in input_ids
+            ]
+        )
+        return input_ids_torch.long(), attention_mask.long().to(self.device), mapping
+
+    def send_responses(self, requests, outputs, mapping):
+        """
+        Scatter method for processing requests and sending responses.
+
+        Args:
+            requests (list): List of Triton InferenceRequest objects.
+            outputs (list): List of output tensors generated by the model.
+            mapping (list): List of indices that map the input tensors to the requests.
+
+        Returns:
+            None
+        """
+        for i in mapping:
+            index = mapping[i]
+            request = requests[index]
+
+            response_sender = request.get_response_sender()
+            # Convert scalar to a one dimensional tensor
+            generated_token = outputs[index][-1].reshape(1)
+
+            self.state[index] = torch.cat([self.state[index], generated_token])
+            if (
+                generated_token.item() == self.tokenizer.eos_token_id
+                or len(self.state[index]) >= self.max_tokens
+            ):
+                flags = pb_utils.TRITONSERVER_RESPONSE_COMPLETE_FINAL
+                self.state[index] = []
+            else:
+                flags = 0
+
+            output_decoded = self.tokenizer.decode(generated_token.cpu().item())
+            response = pb_utils.InferenceResponse(
+                output_tensors=[
+                    pb_utils.Tensor(
+                        "text_output", np.array([output_decoded], dtype=np.object_)
+                    )
+                ]
+            )
+            response_sender.send(response, flags=flags)
+
+    def execute(self, requests):
```

Comment:
The iterative version is really easy to follow - I'm wondering if we can reuse the same state management logic - and just change the execute method to "run to completion" i.e. run until all responses are final? we can remove any requests that have finished from the created batch in a similar fashion as the core would for completed sequences